### PR TITLE
R utilities

### PR DESCRIPTION
**Pull request by Yashwants19**
Hi @coatless @eddelbuettel @rcurtin: I have added all utility functions required by R bindings.
I also added `test_r_binding.R`(hand written binding which contain all the utilities). And I have also add some more test in `test-R_binding.R` which test all the utility functions.

Please guide whether these hand-written binding are correct or not? (Currently we have some more handwritten binding i.e `lars.R` and `pca.R`)
Same in the case of `r_util.cpp` .

If you all agree, Initially we can skip `serialize_to_xml` functionality, as this functionality is not provided by any of the mlpack bindings.

Thank You.

### CODE REVIEW

File: R/matrix_utils.R [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+      char_data <- which(is.na(as.character(as.numeric(x[, 1]))))
+      options(warn = 0)
+      if (length(char_data) != 0) {
+        for (i in 1:length(char_data)) {
+          d[char_data[i]] <- TRUE
+          x[char_data[i], ] <- as.numeric(as.factor(unlist(x[char_data[i], ])))
+        }
+      }
+    }
+    y <- data.matrix(x)
```
Comment (Yashwants19) [inline R/matrix_utils.R]: Hi I think so test is failing here. 
Can we disable this part when `points_are_rows=F`, and can show warning message if user pass`points_are_rows=F` with `matrix_with_info` as : `points_are_rows` will always treated as `TRUE` in case of `data.frame`.?

As this will results into` to_matrix_with_info()` will only accept this type of data.frame: 
``` 
   a  b  c  d  e  f
1  1  2  3  4  5  a
2  6  7  8  9 10  b
3 11 12 13 14 15  c
```
and not this type of data.frame:
```
   a  b  c  d  e
1  1  2  3  4  5
2  6  7  8  9 10
3 11 12 13 14 15
4  a  b  c  d  e 
```
Thank You.

**Comment by eddelbuettel** [comment]
First off, nice work!  But we may want to go a little slower -- PRs with 14 files changed are heavy!

Next, I can confirm that if we `skip()` the test `test_that("TestDataFrameWithInfoColMajor" ...` then the PR passes.  And what happens in that function deserves some discussion.   You create as `y` 

```
R> y <- matrix(c(1:20), ncol = 5)
R> x <- rbind.data.frame(y, "d" = letters[1:5], stringsAsFactors=FALSE)
R> x
  V1 V2 V3 V4 V5
1  1  5  9 13 17
2  2  6 10 14 18
3  3  7 11 15 19
4  4  8 12 16 20
5  a  b  c  d  e
R> 
```

That may "look" ok to a MLPACK user, it does alarming to an R user.  Why?  Look at the class:

```r
R> class(x)
[1] "data.frame"
R> sapply(x, class)
         V1          V2          V3          V4          V5 
"character" "character" "character" "character" "character" 
R> 
```

While `x` is indeed a `data.frame`  as desirect, each column was now converted to a character as that is _the only way to store numbers and characters in each column_ and in R a _data.frame is a column oriented structure_.

In short, you can start from the (numeric) _matrix_ that is `y` and append a single column, but you cannot add a character row across each numeric column.

File: R/matrix_utils.R (hunk 1 above)
Comment (eddelbuettel) [reply R/matrix_utils.R]: Agreed.  See below.  Of these two only the first is sensible in R where _a data.frame is a list of column vectors each of which contains **one** and only one type of data_.

Comment (Yashwants19) [reply R/matrix_utils.R]: I have updated this part.

**Comment by eddelbuettel** [comment]
I haven't had a chance to think this through / look in detail, but are we essentially back to the old discovery of the very first RcppMLPACK "v1" here that ... all R data structures will need to be transposed once before going to MLPACK? 

If so, can we / shall we remove the `points_as_rows` "variable" as it will be a constant? 

Should be think about annotating our R variables to record if they are transposed or not?   (It is very cheap to set an 'attribute' on an R variable to record or store another piece of information.)

**Comment by Yashwants19** [comment]
This sounds good we can remove `points_are_rows` variable , I have no problem with this. :)

>Should be think about annotating our R variables to record if they are transposed or not? (It is very cheap to set an 'attribute' on an R variable to record or store another piece of information.)

I am really sorry I didn't understand this part. If we remove `points_are_rows` , then all R data structure will always be transposed before giving to mlpack and again after getting from mlpack. How this information more useful after removing `points_are_rows`?
I am really sorry, May be I am missing something important here.

**Comment by eddelbuettel** [comment]
Maybe removing `points_are_rows` is the way to go then.

Sorry for being obstruse about the attribute.  Basically R has a function `attr()` 

```r
R> x <- 42
R> attr(x, "raw") <- TRUE
R> x
[1] 42
attr(,"raw")
[1] TRUE
R> 
R> attr(x, "raw")
[1] TRUE
R> 
```

But we don't have to use it. It is simply a way to annotate something.

**Comment by coatless** [comment]
(Some background) 
MLPACK uses matrices in [column-major](https://www.mlpack.org/doc/mlpack-3.2.2/doxygen/matrices.html). However, the meaning of dimensions is swapped from the a statistical modeling perspective. In particular, the paradigm is:

- **Rows** are **features**.
- **Columns** are **observations**.

Quick comparison from one of my lecture slide decks: 
lecture-slide-deck

**Comment by coatless** [comment]
I agree that all `points_as_rows` can be removed in favor of a silent translation or an emphasis within the documentation that matrices should be formatted differently. Worst case scenario, we can set an [environment variable](https://stat.ethz.ch/R-manual/R-patched/library/base/html/EnvVar.html) like `R_TO_MLPACK_TRANPOSE_` to automatically transpose.

That said, I'm not sure two different functions -- `to_matrix()` and `to_matrix_info()` -- are necessary. I think a slight extension to `to_matrix()` that addresses a `character` matrix should be okay.

``` r
to_matrix <- function(x) {
    if (!is.matrix(x) && !is.data.frame(x)) {
        stop("given argument is not matrix-like")
    }
    else if (is.matrix(x) && typeof(x) %in% c("double", "integer")) {
        return(x)
    }
    else if (is.data.frame(x) ) {
        y <- data.matrix(x)
        return(y)
    } else { 
        # Character matrix
        y <- data.matrix(as.data.frame(x))
        return(y)
    }
}

L3 <- LETTERS[1:3]
fac <- sample(L3, 10, replace = TRUE)
(d <- data.frame(x = 1, y = 1:10, fac = fac))
#>    x  y fac
#> 1  1  1   B
#> 2  1  2   C
#> 3  1  3   A
#> 4  1  4   C
#> 5  1  5   A
#> 6  1  6   C
#> 7  1  7   C
#> 8  1  8   C
#> 9  1  9   C
#> 10 1 10   A
to_matrix(d)
#>       x  y fac
#>  [1,] 1  1   2
#>  [2,] 1  2   3
#>  [3,] 1  3   1
#>  [4,] 1  4   3
#>  [5,] 1  5   1
#>  [6,] 1  6   3
#>  [7,] 1  7   3
#>  [8,] 1  8   3
#>  [9,] 1  9   3
#> [10,] 1 10   1

(d2 <- matrix(fac, nrow = 2))
#>      [,1] [,2] [,3] [,4] [,5]
#> [1,] "B"  "A"  "A"  "C"  "C" 
#> [2,] "C"  "C"  "C"  "C"  "A"
to_matrix(d2)
#>      V1 V2 V3 V4 V5
#> [1,]  1  1  1  1  2
#> [2,]  2  2  2  1  1
```

**Comment by eddelbuettel** [comment]
I think we all know. As an aide, I "can't read" you comparison. In "our" standard notation, matrices (the math concept of a 2-d structure) are always n rowx time k columns; rows are always observations, features are always columns.  The are also always contiguous memory; it just depends on which of (i,j) you alter faster.  At work we have this chart:

image

which I find more helpful.  In short a 3x3 matrix is always stored in a vector of length nine.  _Your_ choice or convention if you go row or col order: as in R's.

```r
R> matrix(1:9,3,3,byrow=FALSE)
     [,1] [,2] [,3]
[1,]    1    4    7
[2,]    2    5    8
[3,]    3    6    9
R> matrix(1:9,3,3,byrow=TRUE)
     [,1] [,2] [,3]
[1,]    1    2    3
[2,]    4    5    6
[3,]    7    8    9
R> 
```

The default (for R and others) is the former, as I understand it over it ML space they prefer to transpose their matrices, so out 'tall and skinny' matrix of dim n by k (with n >> k) become k by n: 'wide and shallow'.

(And now I am one comment of yours behind...)

**Comment by eddelbuettel** [comment]
I agree.  A global env var (or what I like better: a logical in a package-wide internal environment; these are easy to set up and I can help; or even an `options()` entry) will do.  

In the discussion above, I think the salient point is that a `data.frame` can have different _column_ types whereas a matrix is only every of one type, often numeric.  As soon as character enters it is all characters in a matrix which is ... pretty useless.   So for one-hot encoding and whatnot, matrices of doubles (or even int if one is sure there are not floats in the data) work, otherwise data.frame is golden and more general.

**Comment by Yashwants19** [comment]
>That said, I'm not sure two different functions -- to_matrix() and to_matrix_info() -- are necessary. I think a slight extension to to_matrix() that addresses a character matrix should be okay.

Hi @coatless both the functions are doing the some specific task i.e:
1) to_matrix: It basically allow the user to pass both matrix and data.frame to the mlpack-bindings. Without this user can only pass matrix to the binding and if user want to pass data.frame, then they them-self have to convert the data.frame to matrix and then pass it to mlpack-bindings (CLI_SetParamMat). 
If you want to remove this -- we can remove it. No problem!
2) to_matrix_with_info: This is use to construct dataset information vector from the given dataset for use in mlpack bindings(CLI_SetParamMatWithInfo). Again here we can pass both `data.frame` as well as `matrix`. 
i) If we only pass matrix, then the vector of info is all FALSE i.e all numeric.
```R
> source('~/RcppMLPACK/R/matrix_utils.R')
> x = matrix(c(1:20), ncol=4)
> to_matrix_with_info(x)
$info
      [,1]  [,2]  [,3]  [,4]
[1,] FALSE FALSE FALSE FALSE

$data
     [,1] [,2] [,3] [,4]
[1,]    1    6   11   16
[2,]    2    7   12   17
[3,]    3    8   13   18
[4,]    4    9   14   19
[5,]    5   10   15   20
````
 ii) but if we pass data.frame, then we construct the vector of info where TRUE represent categorical data, FALSE represent otherwise.
```R
> source('~/RcppMLPACK/R/matrix_utils.R')
> x = matrix(c(1:20), ncol=4)
> y = data.frame(x, letters[1:5])
> y
  X1 X2 X3 X4 letters.1.5.
1  1  6 11 16            a
2  2  7 12 17            b
3  3  8 13 18            c
4  4  9 14 19            d
5  5 10 15 20            e
> to_matrix_with_info(y)
$info
      [,1]  [,2]  [,3]  [,4] [,5]
[1,] FALSE FALSE FALSE FALSE TRUE

$data
     X1 X2 X3 X4 letters.1.5.
[1,]  1  6 11 16            1
[2,]  2  7 12 17            2
[3,]  3  8 13 18            3
[4,]  4  9 14 19            4
[5,]  5 10 15 20            5

```

Thank you.

**Comment by Yashwants19** [comment]
> I agree. A global env var (or what I like better: a logical in a package-wide internal environment; these are easy to set up and I can help; or even an options() entry) will do.

If you can help me with this part, It would be a great help to me. 

Thank You.

**Comment by eddelbuettel** [comment]
> If you can help me with this part, It would be a great help to me.

Absolutely.  And before I go there, let me stress that we _may_ not need it because we just simply "always tranpose" without needing a toggle for it.  

In short, we have three common ways.  Packages can use their own entries in the `options()` to retrieve key-value. A common idiom is the prefix with the package name so `rcppmlpack.transpose` would be a choice.  One can then retrieve with `getOption()` _and very nicely_ set a default if the option was not set:  `val <- getOption("rcppmlpack.transpose", FALSE)` will get you the option value, if set, or FALSE.   

A second approach is to use this, as well as setters and getters, via a package-global variable. A common way is to use a file `R/init.R` and do

```r
.pkgeven <- new.env(parent = emptyenv())

## this function with this signature gets called once at package load
.onLoad <- function(libname, pkgname) {
    ## any code you want there
    .pkgenv[["mykey"]] <- myvalue
}
```
and then have (maybe even non-exported) getters and setters accessing such slots of `.pkgenv`.  I can show you more if need be.

I don't think we need any of it now and here.

File: R/r_util.R [hunk 2]
```diff
@@ -1,6 +1,21 @@
-CLIGetParamMat <- function(paramName, paramValue) {
-  mat <- CLI_GetParamMat(as.character(paramName))
-  nrows <- CLI_GetParamMatRows(as.character(paramName))
-  ncols <- CLI_GetParamMatCols(as.character(paramName))
-  return(matrix(mat, nrows, ncols))
+CLIGetParamMat <- function(param_name, points_are_rows) {
```
Comment (eddelbuettel) [inline R/r_util.R]: These two are the last two uses of `points_are_rows`.  Should we add a default value in the signature to get default behavior without having to set it?

**Review by eddelbuettel — APPROVED** [review]
Still a faily large PR but I think it is good now.  Thanks for making all the changes!

File: R/r_util.R (hunk 2 above)
Comment (coatless) [reply R/r_util.R]: From @eddelbuettel's later comment, the default value would be: 

```r
CLIGetParamMat <- function(param_name, 
                           points_are_rows = getOption('rcppmlpack.tranpose', FALSE)
                           ) { ... }
```

If more are needed, we probably will establish an environment. But, that can be handled at a later time.

**Comment by coatless** [comment]
@Yashwants19 thanks for the detailed walkthrough of both functions. I see the utility for each. I suppose, one idea that I have in my mind is putting the transformation load on `to_matrix()` and having `to_matrix_info()` only be responsible for setting the logical vector on whether a column in the matrix is categorical or not. 

So, something along the lines of: 

```r
# Current definition of to_matrix()
to_matrix = function(x) { ... } 

# Determine column classes
mark_categorical_variable = function(x) {
  d <- sapply(x, class) %in% c("factor", "character", "logical")
  d
}

# Given some matrix-like x (which should be either a matrix or
# data.frame), convert it into a matrix.
to_matrix_with_info <- function(x) {
  
  # Handle transformation
  transformed_x <- to_matrix(x)

  # Figure out categoricals
  info <- mark_categorical_variable(x)
  
  # Return needed data. 
  return(list("info" = info, "data" = transformed_x))
}
```

**Comment by coatless** [comment]
Also, we probably need to have you write a blog entry about GSoC week 1?

File: R/r_util.R (hunk 2 above)
Comment (Yashwants19) [reply R/r_util.R]: I am really sorry. I didn't remove this file in the last commit. But this file is of no use now. As we have decided that we always have to transpose the matrix(before/after), Hence I transferred all the translation to RcppArmadillo because it performed better in microbenchmark for the translation.

**Comment by Yashwants19** [comment]
> @Yashwants19 thanks for the detailed walkthrough of both functions. I see the utility for each. I suppose, one idea that I have in my mind is putting the transformation load on `to_matrix()` and having `to_matrix_info()` only be responsible for setting the logical vector on whether a column in the matrix is categorical or not.
> 
> So, something along the lines of:
> 
> ```r
> # Current definition of to_matrix()
> to_matrix = function(x) { ... } 
> 
> # Determine column classes
> mark_categorical_variable = function(x) {
>   d <- sapply(x, class) %in% c("factor", "character", "logical")
>   d
> }
> 
> # Given some matrix-like x (which should be either a matrix or
> # data.frame), convert it into a matrix.
> to_matrix_with_info <- function(x) {
>   
>   # Handle transformation
>   transformed_x <- to_matrix(x)
> 
>   # Figure out categoricals
>   info <- mark_categorical_variable(x)
>   
>   # Return needed data. 
>   return(list("info" = info, "data" = transformed_x))
> }
> ```

Thank You for the great change. I have updated the function in last commit.

**Comment by Yashwants19** [comment]
> Also, we probably need to have you write a blog entry about GSoC week 1?

I am really sorry about this. I am on it, I am will post this by tomorrow.

**Comment by Yashwants19** [comment]
Hello everyone, Sharing GSoC blog. I am not the best at this part, but I tried to write a blog. Kindly let me know what do you think. You can take a look at this [here](https://medium.com/@yashwantsingh.sngh/gsoc-journey-1st-week-at-mlpack-8574ae3c5530).
Thank You for the great guidance till now :smile:.

**Comment by eddelbuettel** [comment]
Looks great!  

And ... we should probably merge this PR now or am I overlooking any outstanding issues?

**Comment by Yashwants19** [comment]
> Looks great!

Thank You. :)

> And ... we should probably merge this PR now or am I overlooking any outstanding issues?

Should we wait for @rcurtin's input, or merge this PR?

**Comment by eddelbuettel** [comment]
It's always worth waiting for input by @rcurtin, good call.

**Comment by eddelbuettel** [comment]
Paging @rcurtin (and @coatless who hasn't approved?)  ...

**Review by coatless — APPROVED** [review]

**Comment by rcurtin** [comment]
Still coming back up to speed after last week---should have a response shortly once I get everything cached in. :+1:

File: R/pca.R [hunk 3]
```diff
@@ -37,10 +37,6 @@
 #' @param var_to_retain Amount of variance to retain; should be
 #' between 0 and 1.  If 1, all variance is retained.  Overrides -d.  Default
 #' value `0`.
-#' @param copy_all_inputs If specified, all input parameters will be
-#' deep copied before the method is run.  This is useful for debugging
-#' problems where the input parameters are being modified by the algorithm,
-#' but can slow down the code.
```
Comment (rcurtin) [inline R/pca.R:43]: Food for thought---is it worth preserving this option?  Since the mlpack bindings often use std::move() internally, this can mean that if we have some matrix `X` that's used as the training data for some mlpack binding, after we call the mlpack binding, `X` could be empty (due to the internal `std::move()`).  I don't think that we need to make any changes here right now, but we should perhaps keep in mind that we may want to revisit this issue someday and provide that support.

File: R/pca.R [hunk 4]
```diff
@@ -77,7 +72,7 @@ pca <- function(input,
     CLI_SetParamDouble("var_to_retain", var_to_retain)
   }
 
-  if (verbose != FALSE || verbose == TRUE) {
+  if (verbose != FALSE && verbose == TRUE) {
```
Comment (rcurtin) [inline R/pca.R]: Shouldn't just `if(verbose == TRUE)` work here?  (I am not an R expert)

File: R/test_r_binding.R [hunk 5]
```diff
@@ -66,32 +82,85 @@ test_r_binding <- function(double_in,
...
+
+  if (!identical(urow_in, matrix(NA))) {
+    CLI_SetParamURow("urow_in", to_matrix(urow_in))
+  }
+
+  if (!identical(vector_in, NA)) {
+    CLI_SetParamVectorInt("vector_in", vector_in)
+  }
+
+  if (verbose != FALSE && verbose == TRUE) {
```
Comment (rcurtin) [inline R/test_r_binding.R]: Same here, maybe `if (verbose == TRUE)` works?

File: src/RcppExports.cpp [hunk 6]
```diff
@@ -134,15 +134,103 @@ BEGIN_RCPP
     return R_NilValue;
 END_RCPP
 }
+// CLI_SetParamVectorStr
+void CLI_SetParamVectorStr(const std::string& paramName, const std::vector<std::string>& str);
+RcppExport SEXP _RcppMLPACK_CLI_SetParamVectorStr(SEXP paramNameSEXP, SEXP strSEXP) {
+BEGIN_RCPP
+    Rcpp::RNGScope rcpp_rngScope_gen;
```
Comment (rcurtin) [inline src/RcppExports.cpp:141]: This seems unused; did I overlook something?  It seems to be in the other functions too.  (Again, not an R expert, also not an Rcpp expert :))

**Review by rcurtin — APPROVED** [review]
Everything looks good to me. :+1: I don't have enough R knowledge at the moment to run these, but I get the impression that the code here works, and anyway I understand this PR more as a proof-of-concept where we'll get the binding generator to output code like `lars.R`.

I left a few comments, but I don't know how relevant they are, or if I am missing something due to my lack of familiarity with R.  I will come up to speed, but it will take a little while. :)

File: R/pca.R (hunk 4 above)
Comment (eddelbuettel) [reply R/pca.R]: I missed that.  That looks odd.  Even  `if (verbose) { ...`  which is my standard idiom.  It's a logical, no?

File: src/RcppExports.cpp (hunk 6 above)
Comment (eddelbuettel) [reply src/RcppExports.cpp:141]: Machine generated. See top of file.  Ignore macros.

Comment (rcurtin) [reply src/RcppExports.cpp:141]: Ahhh!!  Right.  I missed that.  Thank you!

File: R/pca.R (hunk 4 above)
Comment (Yashwants19) [reply R/pca.R]: I have correct this in the last commit.

File: R/test_r_binding.R (hunk 5 above)
Comment (Yashwants19) [reply R/test_r_binding.R]: Updated.

**Comment by coatless** [comment]
With the `if` statement corrections done, I think we can merge this in. @Yashwants19 would you like to do the honors? 

You may want to "Squash and merge".

**Comment by Yashwants19** [comment]
Thank you for all the guidance. :smile:
